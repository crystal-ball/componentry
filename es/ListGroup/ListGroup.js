import _objectWithoutProperties from "@babel/runtime/helpers/objectWithoutProperties";
import _defineProperty from "@babel/runtime/helpers/defineProperty";

function ownKeys(object, enumerableOnly) { var keys = Object.keys(object); if (Object.getOwnPropertySymbols) { var symbols = Object.getOwnPropertySymbols(object); if (enumerableOnly) symbols = symbols.filter(function (sym) { return Object.getOwnPropertyDescriptor(object, sym).enumerable; }); keys.push.apply(keys, symbols); } return keys; }

function _objectSpread(target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i] != null ? arguments[i] : {}; if (i % 2) { ownKeys(Object(source), true).forEach(function (key) { _defineProperty(target, key, source[key]); }); } else if (Object.getOwnPropertyDescriptors) { Object.defineProperties(target, Object.getOwnPropertyDescriptors(source)); } else { ownKeys(Object(source)).forEach(function (key) { Object.defineProperty(target, key, Object.getOwnPropertyDescriptor(source, key)); }); } } return target; }

import { Children } from 'react';
import { BaseButton } from '../Button/Button';
import elementFactory from '../component-factories/element';
/**
 * ðŸ¤” There are different wrappers for clickable vs non-clickable list groups. (this
 * seems less than ideal Bootstrap, can we always do a div?)
 */

var ListGroup = elementFactory('ListGroup', function (props) {
  var child = Children.toArray(props.children)[0]; // We have to be careful of children like text nodes with this check

  var childProps = child && child.props ? child.props : {};
  return _objectSpread({
    className: 'list-group',
    tag: childProps.href || childProps.onClick ? 'div' : 'ul'
  }, props);
});
var ListGroupItem = elementFactory('ListGroupItem', function (_ref) {
  var active = _ref.active,
      color = _ref.color,
      rest = _objectWithoutProperties(_ref, ["active", "color"]);

  var href = rest.href,
      onClick = rest.onClick;
  return _objectSpread({
    className: _defineProperty({
      'list-group-item': true,
      active: active,
      'list-group-item-action': href || onClick
    }, "list-group-item-".concat(color), color),

    /* eslint-disable no-nested-ternary */
    tag: href || onClick ? href ? 'a' : BaseButton : 'li'
  }, rest);
}); // $FlowFixMe

ListGroup.Item = ListGroupItem;
export default ListGroup;