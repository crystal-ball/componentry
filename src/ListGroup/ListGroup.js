import { Children } from 'react'
import elem from '../elem-factory'
import { useTheme } from '../Theme/Theme'

/**
 * List checks first child (only first child!) to see if it has actions, and if
 * it does, it's not a `li` so we render a `ul`.
 */
export default function ListGroup(props) {
  const child = Children.toArray(props.children)[0]
  // We have to be careful of children like text nodes with this check
  const childProps = child && child.props ? child.props : {}

  return elem({
    as: childProps.href || childProps.onClick ? 'div' : 'ul',
    elemClassName: 'list-group',
    ...useTheme('ListGroup'),
    ...props,
  })
}
ListGroup.displayName = 'ListGroup'

/**
 * List items always have the `list-group-item` class, and items that are
 * actionable (buttons/anchors) have the `list-group-action-item` modifier class
 * added with additional required styles.
 */
ListGroup.Item = function ListGroupItem(props) {
  const { active, color, ...rest } = { ...useTheme('ListGroupItem'), ...props }
  const { href, onClick } = rest

  return elem({
    /* eslint-disable no-nested-ternary */
    as: href || onClick ? (href ? 'a' : 'button') : 'li',
    elemClassName: {
      active,
      disabled: rest.disabled,
      'list-group-item': true,
      'list-group-action-item': href || onClick,
      [`list-group-item-${color}`]: color,
    },
    ...rest,
  })
}
ListGroup.Item.displayName = 'ListGroupItem'
