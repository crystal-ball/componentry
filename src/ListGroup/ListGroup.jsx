// @flow
import { Children } from 'react'

import elementFactory from '../component-factories/element-factory'
import itemFactory from '../component-factories/item-factory'
import type { ElementProps } from '../component-factories/element-factory'

/**
 * ðŸ¤” There are different wrappers for clickable vs non-clickable list groups. (this
 * seems less than ideal Bootstrap, can we always do a div?)
 */

export type Props = ElementProps

const ListGroup = elementFactory({
  name: 'ListGroup',
  classes: 'list-group',
  computedTag: props => {
    const child = Children.toArray(props.children)[0]
    return child && (child.props.href || child.props.onClick) ? 'div' : 'ul'
  },
})

const ListGroupItem = itemFactory({
  name: 'ListGroupItem',
  defaultClasses: 'list-group-item',
  triggerClass: 'list-group-item-action',
})

// $FlowFixMe
ListGroup.Item = ListGroupItem
export default ListGroup
