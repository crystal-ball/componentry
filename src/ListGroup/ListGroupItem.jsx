// @flow
import classNames from 'classnames'

import { BaseButton } from '../Button/Button'
import elementFactory from '../component-factories/element-factory'
import type { ElementProps } from '../component-factories/element-factory'

export type Props = ElementProps

/**
 * List group item of `li`, `a` or `Button`.
 */
export default elementFactory({
  name: 'ListGroupItem',
  clean: ['active'],
  computedClassName: (ctxClassName, propsClassName, { active, href, onClick }) =>
    classNames('list-group-item', ctxClassName, propsClassName, {
      active,
      'list-group-item-action': href || onClick,
    }),
  computedTag: props => {
    // Use BaseButton for onClick elems to b/c we don't want .btn base classes
    if (props.href || props.onClick) return props.href ? 'a' : BaseButton
    // Default to li if not specified or not an action type list
    return 'li'
  },
})
