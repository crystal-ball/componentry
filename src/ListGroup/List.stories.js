/* global module */

import React from 'react'
import { storiesOf } from '@storybook/react'
import { action } from '@storybook/addon-actions'

import ListGroup from './ListGroup'
import Block from '../Block/Block'

const colors = {
  primary: 'Primary',
  secondary: 'Secondary',
  info: 'Info',
  light: 'Light',
  dark: 'Dark',
  success: 'Success',
  warning: 'Warning',
  danger: 'Danger',
}

storiesOf('Display|ListGroup', module)
  .add('<ListGroup />', () => (
    <ListGroup>
      <ListGroup.Item>Radical React</ListGroup.Item>
      <ListGroup.Item>Component Library</ListGroup.Item>
      <ListGroup.Item>Design System</ListGroup.Item>
    </ListGroup>
  ))
  .add('Collection', () => (
    <Block>
      <h2 className='storybook-section-header'>UL list</h2>
      <ListGroup>
        <ListGroup.Item>Radical React</ListGroup.Item>
        <ListGroup.Item>Component Library</ListGroup.Item>
        <ListGroup.Item>Design System</ListGroup.Item>
      </ListGroup>

      <h2 className='storybook-section-header'>Anchor list</h2>
      <ListGroup>
        <ListGroup.Item href='#'>Radical React</ListGroup.Item>
        <ListGroup.Item href='#'>Component Library</ListGroup.Item>
        <ListGroup.Item href='#'>Design System</ListGroup.Item>
      </ListGroup>

      <h2 className='storybook-section-header'>Button list</h2>
      <ListGroup>
        <ListGroup.Item onClick={action('item-click')}>Radical React</ListGroup.Item>
        <ListGroup.Item onClick={action('item-click')}>Component Library</ListGroup.Item>
        <ListGroup.Item onClick={action('item-click')}>Design System</ListGroup.Item>
      </ListGroup>

      <h2 className='storybook-section-header'>List items</h2>
      <ListGroup.Item active>Active list item</ListGroup.Item>
      <ListGroup.Item disabled>Disabled list item</ListGroup.Item>
      {Object.keys(colors).map(color => (
        <ListGroup.Item color={color} key={color}>
          {colors[color]} themed list item
        </ListGroup.Item>
      ))}
    </Block>
  ))
