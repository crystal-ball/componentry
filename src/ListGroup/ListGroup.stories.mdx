import { action } from '@storybook/addon-actions'

import ListGroup from './ListGroup'
import Block from '../Block/Block'
import { colors } from '../storybook-resources'

<Meta title='Components/ListGroup' component={ListGroup} />

<DocsTitle title="List Groups" experimental/>

<ListGroup>
  <ListGroup.Item>Radical React</ListGroup.Item>
  <ListGroup.Item>Component Library</ListGroup.Item>
  <ListGroup.Item>Design System</ListGroup.Item>
</ListGroup>

## Unordered list

<ListGroup>
  <ListGroup.Item>Radical React</ListGroup.Item>
  <ListGroup.Item>Component Library</ListGroup.Item>
  <ListGroup.Item>Design System</ListGroup.Item>
</ListGroup>

## Anchor list

<ListGroup>
  <ListGroup.Item href='#'>Radical React</ListGroup.Item>
  <ListGroup.Item href='#'>Component Library</ListGroup.Item>
  <ListGroup.Item href='#'>Design System</ListGroup.Item>
</ListGroup>

## Button list

<ListGroup>
  <ListGroup.Item onClick={action('item-click')}>Radical React</ListGroup.Item>
  <ListGroup.Item onClick={action('item-click')}>Component Library</ListGroup.Item>
  <ListGroup.Item onClick={action('item-click')}>Design System</ListGroup.Item>
</ListGroup>

## List items

<ListGroup>
  <ListGroup.Item active>Active list item</ListGroup.Item>
  <ListGroup.Item disabled>Disabled list item</ListGroup.Item>
  {colors.map(color => (
    <ListGroup.Item color={color} key={color}>
      {colors[color]} themed list item
    </ListGroup.Item>
  ))}
</ListGroup>
