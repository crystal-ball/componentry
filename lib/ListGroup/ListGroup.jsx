import React, { Children } from 'react'
import { func, node, oneOfType } from 'prop-types'

import ListGroupItem from './ListGroupItem'
import elementFactory from '../utils/element-factory'

const Container = elementFactory({ classes: 'list-group' })

ListGroup.Item = ListGroupItem

ListGroup.propTypes = {
  As: oneOfType([func, node]),
  children: node
}

ListGroup.defaultProps = {
  As: null,
  children: null
}

/**
 * To Document:
 * - There are different wrappers for clickable vs non-clickable list groups. (this
 *   seems less than ideal Bootstrap, can we always do a div?)
 * - Only the first child is checked for an href or onClick for perf. If it is variable,
 *   (which is probably not good in the first place), pass a specific As.
 */
export default function ListGroup({ As, ...rest }) {
  let children = rest.children

  // If As has been configured, use it regardless
  if (!As && children) {
    children = Children.toArray(children)
    if (children.length) {
      const { href, onClick } = children[0].props
      // If children have an href or onClick, we need a div wrapper b/c children will
      // be either <button> or <a> elements and not <li> elements
      As = href || onClick ? 'div' : 'ul'
    } else {
      As = 'div'
    }
  } else {
    As = As || 'ul'
  }

  return <Container As={As} {...rest} />
}
