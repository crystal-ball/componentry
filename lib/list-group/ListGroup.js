import React, { Component, Children } from 'react';
import { any, node, string } from 'prop-types';

import ListGroupItem from './ListGroupItem';
import classNames from '../utils/classnames';

/**
 * To Document:
 * - There are different wrappers for clickable vs non-clickable list groups. (this
 *   seems less than ideal Bootstrap, can we always do a div?)
 * - Only the first child is checked for an href or onClick for perf. If it is variable,
 *   (which is probably not good in the first place), pass a specific As.
 */
export default class ListGroup extends Component {
  static Item = ListGroupItem

  static propTypes = {
    As: any,
    children: node,
    className: string,
  }

  static defaultProps = {
    As: null,
    children: null,
    className: '',
  }

  render() {
    let {
      As,
      children,
      className,
      ...other
    } = this.props;

    className = classNames('list-group', className);

    // If As has been configured, use it regardless
    if (!As && children) {
      let { href, onClick } = Children.toArray(children)[0].props;
      // If children have an href or onClick, we need a div wrapper b/c children will
      // be either <button> or <a> elements and not <li> elements
      As = href || onClick ? 'div' : 'ul';
    } else {
      As = As || 'ul';
    }

    return(
      <As className={className} {...other}>
        {children}
      </As>
    );
  }
}
