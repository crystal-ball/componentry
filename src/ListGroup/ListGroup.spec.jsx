import React from 'react'
import { shallow } from 'enzyme'
import renderer from 'react-test-renderer'

import ListGroup from './ListGroup'
import elementTests from '../utils-test/element-tests'

describe('<ListGroup />', () => {
  // Basic library element test suite
  elementTests(ListGroup)

  test('should render a container ul with class list-group by default', () => {
    const wrapper = shallow(<ListGroup />)
    expect(wrapper.find('ul').length).toEqual(1)
    expect(wrapper.find('.list-group').length).toEqual(1)
  })

  test('should render a div when children have onclick', () => {
    const wrapper = shallow(
      <ListGroup>
        <ListGroup.Item onClick={() => {}} />
      </ListGroup>,
    )
    expect(wrapper.find('ul').length).toBeFalsy()
    expect(wrapper.find('div').length).toBeTruthy()
  })

  test('should render a div when children have hrefs', () => {
    const wrapper = shallow(
      <ListGroup>
        <ListGroup.Item href="test" />
      </ListGroup>,
    )
    expect(wrapper.find('ul').length).toBeFalsy()
    expect(wrapper.find('div').length).toBeTruthy()
  })
})

// Snapshots
// ---------------------------------------------------------------------------
describe('<ListGroup /> snapshots', () => {
  it('renders correctly', () => {
    const tree = renderer
      .create(
        <ListGroup>
          <ListGroup.Item>Item 1</ListGroup.Item>
          <ListGroup.Item>Item 2</ListGroup.Item>
        </ListGroup>,
      )
      .toJSON()
    expect(tree).toMatchSnapshot()
  })
})
