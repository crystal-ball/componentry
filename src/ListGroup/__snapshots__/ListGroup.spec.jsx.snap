// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ListGroup /> snapshots renders correctly 1`] = `
<ul
  className="list-group"
>
  <li
    className="list-group-item"
  >
    Item 1
  </li>
  <li
    className="list-group-item"
  >
    Item 2
  </li>
</ul>
`;
